File: src/main/java/engine/rendering/instances/postprocessing/mix/MixPostProcessor.java
package engine.rendering.instances.postprocessing.mix;

import engine.rendering.RenderOutputData;
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor;
import engine.rendering.abstracted.postprocessing.PostProcessor;
import engine.utils.libraryBindings.opengl.shaders.RenderState;
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty;
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty;
import engine.utils.libraryBindings.opengl.textures.ITexture;
import engine.utils.libraryBindings.opengl.textures.Texture;

public class MixPostProcessor extends AbstractPostProcessor {

  private static final String FRAG_FILE = "/shaders/postprocessing/mix/MixFragment.glsl";

  private float scalar = 0.5f;
  private Texture combination = Texture.NONE;

  public MixPostProcessor() throws Exception {
    super(FRAG_FILE);
    initialize();
  }

  public MixPostProcessor(float scalar) throws Exception {
    this();
    this.scalar = scalar;
  }

  public MixPostProcessor(Texture combination) throws Exception {
    this();
    this.combination = combination;
  }

  private void initialize() {
    getShadersProgram().addPerRenderUniform(new UniformTextureProperty<RenderOutputData>("texture1", 0) {
      @Override
      public ITexture getUniformValue(RenderState<RenderOutputData> state) {
        return state.getInstance().colour;
      }
    });

    getShadersProgram().addPerRenderUniform(new UniformTextureProperty<RenderOutputData>("texture2", 1) {
      @Override
      public ITexture getUniformValue(RenderState<RenderOutputData> state) {
        return combination;
      }
    });

    getShadersProgram().addPerRenderUniform(new UniformFloatProperty<RenderOutputData>("scalar") {
      @Override
      public float getUniformValue(RenderState<RenderOutputData> state) {
        return scalar;
      }
    });
  }
}

File: src/main/java/engine/rendering/instances/postprocessing/mix/MixPostProcessor.kt
package engine.rendering.instances.postprocessing.mix

import engine.rendering.RenderOutputData
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor
import engine.rendering.abstracted.postprocessing.PostProcessor
import engine.utils.libraryBindings.opengl.shaders.RenderState
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty
import engine.utils.libraryBindings.opengl.textures.ITexture
import engine.utils.libraryBindings.opengl.textures.Texture

class MixPostProcessor() : AbstractPostProcessor(FRAG_FILE), PostProcessor {

    constructor(scalar: Float) : this() {
        this.scalar = scalar
    }

    constructor(combination: Texture) : this() {
        this.combination = combination
    }

    var scalar: Float = .5f
    var combination: Texture = Texture.NONE

    init {
        shadersProgram.addPerRenderUniform(object : UniformTextureProperty<RenderOutputData>("texture1", 0) {
            override fun getUniformValue(state: RenderState<RenderOutputData>): ITexture {
                return state.instance.colour
            }
        })
        shadersProgram.addPerRenderUniform(object : UniformTextureProperty<RenderOutputData>("texture2", 1) {
            override fun getUniformValue(state: RenderState<RenderOutputData>): ITexture {
                return combination
            }
        })
        shadersProgram.addPerRenderUniform(object : UniformFloatProperty<RenderOutputData>("scalar") {
            override fun getUniformValue(state: RenderState<RenderOutputData>): Float {
                return scalar
            }
        })
    }

    companion object {
        private const val FRAG_FILE = "/shaders/postprocessing/mix/MixFragment.glsl"
    }
}
